doc/reference/trd-appid.md

```diff
@@ -0,0 +1,396 @@
+hash of the application binary in this Tock binary, encrypted by the private key 
+of the public key in the header.
+
+The `Rsa3072KeyWithID` value has a data of length of 768 bytes. It contains a public 3072-bit
+RSA key (384 bytes), followed by a 384-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+The `Rsa4096KeyWithID` value has a data of length of 1024 bytes. It contains a public 4096-bit
+RSA key (512 bytes), followed by a 512-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+
+4 `Verifier` trait
+===============================
+
+The `Verifier` trait defines an interface to a module that accepts,
+passes on, or rejects application credentials. When a Tock board
+asks the kernel to load processes, it passes a reference to a 
+`Verifier`, which the kernel uses to check credentials.
+
+
+```rust
+pub enum VerificationResult {
+  Accept,
+  Pass,
+  Reject
+}
+
+pub trait Verifier {
+  fn require_credential(&self) -> bool;
+  fn check_credentials(&self, 
+                       credentials: &TbfHeaderV2Credentials, 
+                       binary: &mut [u8]) -> VerificationResult;
+}
+```
+
+The kernel, when it loads a Tock binary, scans its headers in order
+from the beginning of the Tock binary. At each
+`TbfHeaderV2Credentials` header it encounters, it calls
+`check_credentials` on the provided `Verifier`. If the `Verifier`
+returns `Accept`, the kernel stops processing credentials and
+continues loading the Tock binary. If the `Verifier` returns `Reject`,
+the kernel stops processing credentials and terminates loading the
+Tock binary. If the `Verifier` returns `Pass`, the kernel tries the
+next `TbfHeaderV2Credentials`, if there is one. 
+
+If the kernel reaches the end of the TBF headers without encountering
+a `Reject` or `Accept` result, it calls `require_credentials` to ask
+the `Verifier` what the default behavior is.  If `require_credentials`
+returns `true`, the kernel rejects the Tock binary and terminates
+loading it. If `require_credentials` returns `false`, the kernel
+accepts the Tock binary and continues loading it. If a Tock binary has
+no `TbfHeaderV2Credentials` headers then there will be no `Accept` or
+`Reject` results and `require_credentials` defines whether to load
+such a binary.
+
+An implementer of `Verifier` sets the security policy of Tock binary
+loading by deciding which types of credentials, and which credentials,
+are acceptable and which are rejected.
+
+If `check_credentials` returns `Accept` for a
+`TbfHeaderV2Credentials`, the kernel stores a reference to this
+`TbfHeaderV2Credentials` in the process structure. This data
+represents the acting credentials of the process.
```

Comment:
If the kernel stores a reference to the `TbfHeaderV2Credentials` rather than an application ID, how does the kernel access the application ID?

Reply:
My thought is it would store both.

Reply:
What decodes the `TbfHeaderV2Credentials` to find the application ID? It seems like the `Verifier` would know how to do that, but not the core kernel.

Reply:
The `Compress` trait does it. The intention (I think the TRD talks about this) is that the same structure implements `Verifier` and `Compress`.

Reply:
`Compress` doesn't give a way to access the ID either. E.g. how would an IPC capsule access the app ID for a process that `to_short_id` returns `None` for?

Reply:
It wouldn't! So this is a good question: is it that there is always a short ID?

I am a little worried about collisions when there are some IDs generated one way and other IDs generated another way. Unless we partition the ID space, or if both algorithms inspect the current IDs to ensure there is no collision (they can't be a deterministic mapping from only credentials), there might be a collision.

Reply:
I was under the impression that there will always be a short ID and it's up to the `Compress` trait to ensure they don't collide

Reply:
"It's up to the Compress trait" means that they must walk the processes...

Reply:
It doesn't have to. A simple counter for example wouldn't have to worry about walking the processes. But some sort of hash would probably need to make sure that there isn't a collision

Reply:
To ensure uniqueness, it has to store state. If the IDs are persistent then the state must depend on the Tock binary in some way. One could possibly store persistent IDs locally in the `Compress` module, but this would require monitoring process end and a table sized to NUM_PROCESSES.